### ISSUE TITLE
broken in 2.1.41

### ISSUE THREAD

**Issue by ijgnd** [ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style#45]
I can no longer apply classes to selected text: the html source code of the field is unchanged.

I tried without any other add-ons, the custom styles version I used was the latest commit from this repo. It worked in 2.1.40. I think the problems is the javascript code used in the add-on.

@hgiesel: Since you know js and made some changes to the editor in 2.1.41 maybe you could have a look at this? Thanks.

**Comment by hgiesel** [comment]
Yep, most certainly. I'm currently in the process of updating add-ons for 2.1.41. I will have a look next week.

**Comment by ijgnd** [comment]
that's great. thanks.

**Comment by hgiesel** [comment]
I've updated it. You can apply spans again in the editor, and the styling will be visible.

Rangy will no longer work with the new editor code, so it will always use the "text" solution. The functionality of modifying the Editor CSS might be become a future Anki feature: https://github.com/ankitects/anki/pull/1049.

Let me know if you find any more issues, @ijgnd.

**Comment by ijgnd** [comment]
Thank you very much for updating to the new version.

In general it mostly works again. It doesn't work for me if the selection covers multiple paragraphs/divs. I tested in a new profile with only the custom styles add-on from addons.ankiweb with the binary 41beta6 for linux.

&nbsp;

Example: When I have this in my field:

    <div>one</div><div>two</div>

and I select the last letter of "one" and the first o "two" so that my selection is among two divs and I tell the add-on to apply a class I get the following source code:

    <div>on</div><span class="bg_green"><div>e</div><div>t</div></span><div>wo</div>

no text is shown that has the styling from the class (because there's divs inside this span?)

**Comment by ijgnd** [comment]
Another problem is that now line breaks are added.

&nbsp;

Same example as before: When I have this in my field:

    <div>one</div><div>two</div>

and I select the last letter of "one" and the first o "two" so that my selection is among two divs and I tell the add-on to apply a class with the old version/rangy it was rendered into two lines:

    one
    two
    
The modified source produced by the old version with rangy was `<div>on<span class="bg_green">e</span></div><div><span class="bg_green">t</span>wo</div>`.

Now with the most recent version from 2021-03-03 after applying the highlighting it's rendered into four lines like this:

    on
    e
    t
    wo

&nbsp;

I guess that's the result of "Rangy will no longer work"/"Do not use rangy -> it's not compatible with the editor window"?

You still load rangy into the webview (I mean [this](https://github.com/ijgnd/anki__editor__apply__font_color__background_color__custom_class__custom_style/blob/8e951fd06ea0e2ee7b30af400736d92772bdade0/src/webview/editor_set_css_js_for_webview.py#L134)). Does this mean that in theory it could be adjusted to work again and maybe someone in the future will adapt it? I'll have to learn more html/js before I could investigate this which won't happen before the end of the summer. Or is there really no chance that rangy (that hasn't been changed in the last six years) can work due to your recent changes to the anki editor which now uses more recent browser technologies?
